### ISSUE TITLE
[FEAT] Project idea

### ISSUE THREAD

**Issue by Varun-Kolanu** [Varun-Kolanu/csoc_flutter#1]
**Is your feature request related to a problem? Please describe.**
The idea for the project to choose is not yet decided. We could have chosen it, but as this project is everyone's, you have liberty to suggest ideas. 

**Describe the solution you'd like**
Suggest an idea by 30th May EOD to work on the following 4 weeks. The idea should be useful for everyone and unique.

**Describe alternatives you've considered**
If none of the solutions are feasible or useful, we will decide any new idea that everyone can follow and is unique.

**Comment by saikat-iit** [comment]
**Idea Proposal: Music Download App**

**Problem Statement:**
- Difficulty in extracting music from platforms like Instagram Reels, YouTube, etc.
- Current solutions like Shazam are not always effective, especially with obscure or unreleased tracks.
- Users face inconvenience and frustration when searching for song names in comments, often encountering fake names.

**Solution Overview:**
- Develop an app that allows users to download music from any URL, including Instagram Reels, YouTube, and other platforms.
- Unlike Shazam, which identifies songs, this app will download the original track associated with the URL.
- Provide users with the source information of the downloaded track for transparency and credibility.

**Key Features:**
- URL-based music download: Users can input URLs from various platforms to extract the music.
- Original track download: The app retrieves the full original track rather than extracting audio from videos.
- Source information: Display the source of the downloaded track to ensure authenticity and reliability.

**Comment by devrungta** [comment]
Social Media App
**Features**:-

- _User_ Authentication_

Sign Up / Login: Safe registration and login.
Logout: Safe logout function.

- _User Profiles_

Profile Management: Users can update their information selectively like bio and pfp and all, but not birthdate and gender and other things.
View Profiles: Users can view other peoples' profiles.

-_Posts_
Create Post: Users can post text, images, and videos.
Edit/Delete Post: Users can modify or remove their posts.

-_Feed_
Timeline Feed: Display posts from followed users.
Like, Comment, Share and Save functions.

-_Notifications_
Real-Time Notifications: Notifications about activities likes, comments, and follows.

-_Search and Follow_
Search Users: Users can find other users using username or real name. The ones closer geographically should come up first (for same names)
Follow/Unfollow: Users can follow or unfollow others.

-_DMing_
One-on-One Chat: Real-time messaging.
Group Chat

-_Advanced Features_
Story Feature
Create and View Stories

-_Live Streaming_
This will be tough I think

-_Advanced Search_
Filter by Hashtags and Locations

**Implementation Plan**

- Project Setup: Create a new Flutter project and set up Firebase.
- Authentication: Implement registration and login with Firebase Auth.
- Profile Management: Develop profile features.
- Posts and Feed: Create UI for posts and display them in a feed.
- Notifications: Use Firebase Cloud Messaging for notifications.
- Search and Follow: Implement user search and follow features.
- Direct Messaging: Develop real-time chat.
- Advanced Features: Add stories, live streaming, advanced search, notifications.

**Evaluation Criteria**

- Functionality: Are features working as expected?
- Code Quality: Is the code clean and maintainable?
- UI/UX Design: How user-friendly and visually appealing is the app?
- State Management: Effective use of state management.
- Performance: Is the app responsive?
- Innovation: Any creative features beyond the basics?
`Basically Instagram`

**Comment by Varun-Kolanu** [comment]
> Social Media App **Features**:-
> 
> * _User_ Authentication_
>   Sign Up / Login: Safe registration and login.
>   Logout: Safe logout function.
>   Also remember the IP or the google account on the phone, or use tokens, so we don't have to login again and again
> * _User Profiles_
>   Profile Management: Users can update their information selectively like bio and pfp and all, but not birthdate and gender and other things.
>   View Profiles: Users can view other peoples' profiles.
> * _Posts_
>   Create Post: Users can post text, images, and videos.
>   Edit/Delete Post: Users can modify or remove their posts.
> * _Feed_
>   Timeline Feed: Display posts from followed users.
>   Like, Comment, Share and Save functions.
> * _Notifications_
>   Real-Time Notifications: Notifications about activities likes, comments, and follows.
> * _Search and Follow_
>   Search Users: Users can find other users using username or real name. The ones closer geographically should come up first (for same names)
>   Follow/Unfollow: Users can follow or unfollow others.
> * _DMing_
>   One-on-One Chat: Real-time messaging.
>   Group Chat
> * _Advanced Features_
>   Story Feature
>   Create and View Stories
> * _Live Streaming_
>   This will be tough I think
> * _Advanced Search_
>   Filter by Hashtags and Locations
> 
> **Implementation Plan**
> 
> * Project Setup: Create a new Flutter project and set up Firebase.
> * Authentication: Implement registration and login with Firebase Auth.
> * Profile Management: Develop profile features.
> * Posts and Feed: Create UI for posts and display them in a feed.
> * Notifications: Use Firebase Cloud Messaging for notifications.
> * Search and Follow: Implement user search and follow features.
> * Direct Messaging: Develop real-time chat.
> * Advanced Features: Add stories, live streaming, advanced search, notifications.
> 
> **Evaluation Criteria**
> 
> * Functionality: Are features working as expected?
> * Code Quality: Is the code clean and maintainable?
> * UI/UX Design: How user-friendly and visually appealing is the app?
> * State Management: Effective use of state management.
> * Performance: Is the app responsive?
> * Innovation: Any creative features beyond the basics?
>   `Basically Instagram with less features`

Nice idea for learning.
But it would be better if we implement something unique which no one has ever built or failed to build. Social media app is available in many youtube videos, and it has become a cliche tbh. Think like this ki, instagram is already available on playstore. Why would any user use ours when an authentic insta exists from Meta already...

**Comment by criticic** [comment]
For the music downloader app we should probably use something like [yt-dlp](https://github.com/yt-dlp/yt-dlp), as scraping instagram, and also avoid ratelimits and other protections would prolly be [pretty hard](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/instagram.py)

**Comment by Varun-Kolanu** [comment]
> For the music downloader app we should probably use something like [yt-dlp](https://github.com/yt-dlp/yt-dlp), as scraping instagram, and also avoid ratelimits and other protections would prolly be [pretty hard](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/instagram.py)
> 
> also how will we search the full track?

Yeah it seems to be hard. This should be the app which anyone can follow from beginners to advanced programmers. 

And it's not restricted to give only 1 idea. You can give as many ideas as you can. We'll select the best and feasible one

**Comment by saikat-iit** [comment]
> > For the music downloader app we should probably use something like [yt-dlp](https://github.com/yt-dlp/yt-dlp), as scraping instagram, and also avoid ratelimits and other protections would prolly be [pretty hard](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/instagram.py)
> > 
> > also how will we search the full track?
> 
> Yeah it seems to be hard. This should be the app which anyone can follow from beginners to advanced programmers. 
> For How to part, let's ask if @saikat-iit has any implementation he thought of.. 
> 
> And it's not restricted to give only 1 idea. You can give as many ideas as you can. We'll select the best and feasible one

Actually I don't have a solid Idea on how to achieve this right now. But I was thinking about using concept of Audio Fingerprint and Open Source Libraries like [EchoPrint](https://github.com/spotify/echoprint-codegen). But unfortunately the Library is not getting maintained, So I don't know if this might work, as EchoPrint uses it's own Database, and I don't know if Spotify API have the concept of Audio Fingerprint. 

Come to think of it, it's not a good beginner's project, as I am unsure about it's success rate.

I'll think about something else which is unique that has real life usage.

**Comment by criticic** [comment]
A couple of ideas off the top of my head, not really that unique but here:
## 1) Something like https://anilist.co/ but for Movies/Shows/Books
- I would really love to use something like this to track all the movies I have ever watched, I cant find a good place (with a simple ui and less clicks) where it shows me not only the list of movies, but also the total watch time (in days), or watche over the years or somehting
- would utilize a lot of things used in apps, grids, imageviews, layouts, and would also prolly use an api like tmdb
- Could find lots of intersting designs on dribbble

## 2) Expense Manager
- similar to the now deprecated salt passbook app
- one would enter his expenses, and category and all, and would be shown spending across various categories in a month year etc
- (no idea how It could be implemented) automatically categorize from just text input like the salt app (so writing "canteen maggi 35" would automatically add a +35 expense with name "canteen maggi" and categorize it as food)
Cons
- Without the automatic categorization feature, it would be way too manual, and would basically be like a upgraded todo app

## 3) Complicated Alarm Clock
- Basically we could have various mechanisms to stop the alarm, like solve a math problem, or just some task like taking photos (of something) using the camera, shouting in the microphone etc

**Comment by saikat-iit** [comment]
> A couple of ideas off the top of my head, not really that unique but here:
> ## 1) Something like https://anilist.co/ but for Movies/Shows/Books
> - I would really love to use something like this to track all the movies I have ever watched, I cant find a good place (with a simple ui and less clicks) where it shows me not only the list of movies, but also the total watch time (in days), or watche over the years or somehting
> - would utilize a lot of things used in apps, grids, imageviews, layouts, and would also prolly use an api like tmdb
> - Could find lots of intersting designs on dribbble
> 
> ## 2) Expense Manager
> - similar to the now deprecated salt passbook app
> - one would enter his expenses, and category and all, and would be shown spending across various categories in a month year etc
> - (no idea how It could be implemented) automatically categorize from just text input like the salt app (so writing "canteen maggi 35" would automatically add a +35 expense with name "canteen maggi" and categorize it as food)
> Cons
> - Without the automatic categorization feature, it would be way too manual, and would basically be like a upgraded todo app
> 
> ## 3) Complicated Alarm Clock
> - Basically we could have various mechanisms to stop the alarm, like solve a math problem, or just some task like taking photos (of something) using the camera, shouting in the microphone etc

I liked the concept of Complex Alarm.

**Comment by Varun-Kolanu** [comment]
> A couple of ideas off the top of my head, not really that unique but here:
> 
> ## 1) Something like https://anilist.co/ but for Movies/Shows/Books
> * I would really love to use something like this to track all the movies I have ever watched, I cant find a good place (with a simple ui and less clicks) where it shows me not only the list of movies, but also the total watch time (in days), or watche over the years or somehting
> * would utilize a lot of things used in apps, grids, imageviews, layouts, and would also prolly use an api like tmdb
> * Could find lots of intersting designs on dribbble
> 
> ## 2) Expense Manager
> * similar to the now deprecated salt passbook app
> * one would enter his expenses, and category and all, and would be shown spending across various categories in a month year etc
> * (no idea how It could be implemented) automatically categorize from just text input like the salt app (so writing "canteen maggi 35" would automatically add a +35 expense with name "canteen maggi" and categorize it as food)
>   Cons
> * Without the automatic categorization feature, it would be way too manual, and would basically be like a upgraded todo app
> 
> ## 3) Complicated Alarm Clock
> * Basically we could have various mechanisms to stop the alarm, like solve a math problem, or just some task like taking photos (of something) using the camera, shouting in the microphone etc

Hmm cool. 
Anilist is nice.
Expense manager is also good with that automatic categorization. 
Alarm clock wala Alarmy app me already hai na. Ye mera roommate use karta tha with tha and mujhe deta tha utke to solve and wo so jaata 😂

**Comment by devrungta** [comment]
> Meta

I hoped to give simple idea and I actually forgot about the YT tutorials for app dev...
I thought this would be easy and test a lot of things...

**Comment by devrungta** [comment]
> > A couple of ideas off the top of my head, not really that unique but here:
> > ## 1) Something like https://anilist.co/ but for Movies/Shows/Books
> > 
> > * I would really love to use something like this to track all the movies I have ever watched, I cant find a good place (with a simple ui and less clicks) where it shows me not only the list of movies, but also the total watch time (in days), or watche over the years or somehting
> > * would utilize a lot of things used in apps, grids, imageviews, layouts, and would also prolly use an api like tmdb
> > * Could find lots of intersting designs on dribbble
> > 
> > ## 2) Expense Manager
> > 
> > * similar to the now deprecated salt passbook app
> > * one would enter his expenses, and category and all, and would be shown spending across various categories in a month year etc
> > * (no idea how It could be implemented) automatically categorize from just text input like the salt app (so writing "canteen maggi 35" would automatically add a +35 expense with name "canteen maggi" and categorize it as food)
> >   Cons
> > * Without the automatic categorization feature, it would be way too manual, and would basically be like a upgraded todo app
> > 
> > ## 3) Complicated Alarm Clock
> > 
> > * Basically we could have various mechanisms to stop the alarm, like solve a math problem, or just some task like taking photos (of something) using the camera, shouting in the microphone etc
> 
> I liked the concept of Complex Alarm.
> 
> Instead of Maths Problem (as it will restrict no. of users) (what if the person don't know maths), Using ReCAPTCHA v1 to stop the alarm will be fun 😂.
> 
> **Combined Idea Proposal: Captcha Alarm**
> 
> **Concept:** A alarm system that integrates Captcha technology to ensure users actively engage with the alarm to turn it off.
> 
> **Features:**
> 
> * **Captcha Integration:**
>   
>   * Requires users to solve a Recaptcha V1 puzzle to deactivate the alarm.
> * **Enhanced Engagement:**
>   
>   * Promotes cognitive engagement by requiring users to solve a Captcha, ensuring they are fully awake and alert.
> * **Security:**
>   
>   * Adds an extra layer of security by verifying the user's humanity before allowing them to silence the alarm.
We could give them multiple options to try and give them more ways to disarm the alarm.
or, we could use the GPS system of the phone and disarm the alarm only when the user has moved a certain distance that they choose with a minimum distance being set by us as an option.
we could also apply a resistance for VPN.
We could make it so the alarm keeps getting louder.
A feedback system for how much you have walked in a day, week, month and time charts on how long you take for the task, speed charts and some more analytics to give the user an idea of their workout.

**Comment by devrungta** [comment]
**Proposal**
A special calendar app for lazy or forgetful people.

**Features**

- User Profile 

- Weekly Schedule
 Set class schedule and other weekly engagements (e.g., club activities, dates)

- Attendance Tracking
 Mark attendance daily as absent, present, proxied, or canceled for all activities

- Attendance Statistics
 Track how many classes/activities attended and calculate how many can be bunked
 Provide two statistics: based on real attendance and proxied attendance

- Priority Setting
Assign priorities to activities (e.g., must go, leave it, chill hai), like for manuprac, you can't skip

- Social Integration
 Connect with friends to compare attendance for common things and other contacts for more details like b'days and all

- Absence Management
 Input reasons for absences and generate apology emails or WhatsApp messages with preset patterns

- Notifications and Reminders
 Receive reminders and notifications for scheduled activities whenever user asks

**Implementation Plan**

- Project Setup
- User Profile
- Weekly Schedule
- Attendance Tracking
- Attendance Statistics
- Priority Setting
- Social Integration
- Absence Management
- Notifications and Reminders

**Evaluation Criteria**

- Functionality
- Code Quality
- UI/UX Design
- State Management
- Testing
- Performance
- Innovation

**Comment by Varun-Kolanu** [comment]
@devrungta Hehe nice one actually. I forget almost everything 😭

In attendance tracking, I am too lazy to open the app and mark attendance. So I deleted the app eventually 😅
For this, the following can be done:
1. There should be a time, which user choses. At that time daily on working days, app sends a notification to mark attendance. 
2. We can put a feature that user can save his classroom location. So, whenever he comes inside some radius, with class as center, app sends a notification whether to mark attendance for that class. 

It might be so repetitive to create the time table of the class. So, in the Social Integration section we can add the following features:
1. User can see the friends' time table created. Then there will be a button to copy the same time table.
2. A user can choose who can see their attendance: Friends, Everyone, No one
3. Generally, if we forget to mark whether class was cancelled we will not be able to remember what happened. So, the cancellation of classes will be visible to the friends no matter what, because no privacy is getting disturbed in that. Then, a friend can see the class cancellations from other friends and mark the same in his.
4. We can request our friends to mark proxy. Friends can accept or reject the request. Only one user can accept the proxy request. When a friend has accepted a proxy request, all the other requests to friends will be revoked. Then the accepted friend can mark whether he was successful in marking proxy or not. Based on that, the attendance of that day will be automatically marked as proxied or absent in the user's time table.

**Comment by saikat-iit** [comment]
> @devrungta Hehe nice one actually. I forget almost everything 😭
> 
> In attendance tracking, I am too lazy to open the app and mark attendance. So I deleted the app eventually 😅
> For this, the following can be done:
> 1. There should be a time, which user choses. At that time daily on working days, app sends a notification to mark attendance. 
> 2. We can put a feature that user can save his classroom location. So, whenever he comes inside some radius, with class as center, app sends a notification whether to mark attendance for that class. 
> 
> It might be so repetitive to create the time table of the class. So, in the Social Integration section we can add the following features:
> 1. User can see the friends' time table created. Then there will be a button to copy the same time table.
> 2. A user can choose who can see their attendance: Friends, Everyone, No one
> 3. Generally, if we forget to mark whether class was cancelled we will not be able to remember what happened. So, the cancellation of classes will be visible to the friends no matter what, because no privacy is getting disturbed in that. Then, a friend can see the class cancellations from other friends and mark the same in his.
> 4. We can request our friends to mark proxy. Friends can accept or reject the request. Only one user can accept the proxy request. When a friend has accepted a proxy request, all the other requests to friends will be revoked. Then the accepted friend can mark whether he was successful in marking proxy or not. Based on that, the attendance of that day will be automatically marked as proxied or absent in the user's time table.

We can also think about some algorithm to give approx % of attendance (if possible)

Create a Database of colleges, year, course to get approx amount classes taken and then we can provide approx no. of classes he/she can miss (also % of attendance) (there will be 2 % of attendance, one with considering proxy and one without proxy, just to be safe)

**Comment by devrungta** [comment]
> > @devrungta Hehe nice one actually. I forget almost everything 😭
> > In attendance tracking, I am too lazy to open the app and mark attendance. So I deleted the app eventually 😅
> > For this, the following can be done:
> > 
> > 1. There should be a time, which user choses. At that time daily on working days, app sends a notification to mark attendance.
> > 2. We can put a feature that user can save his classroom location. So, whenever he comes inside some radius, with class as center, app sends a notification whether to mark attendance for that class.
> > 
> > It might be so repetitive to create the time table of the class. So, in the Social Integration section we can add the following features:
> > 
> > 1. User can see the friends' time table created. Then there will be a button to copy the same time table.
> > 2. A user can choose who can see their attendance: Friends, Everyone, No one
> > 3. Generally, if we forget to mark whether class was cancelled we will not be able to remember what happened. So, the cancellation of classes will be visible to the friends no matter what, because no privacy is getting disturbed in that. Then, a friend can see the class cancellations from other friends and mark the same in his.
> > 4. We can request our friends to mark proxy. Friends can accept or reject the request. Only one user can accept the proxy request. When a friend has accepted a proxy request, all the other requests to friends will be revoked. Then the accepted friend can mark whether he was successful in marking proxy or not. Based on that, the attendance of that day will be automatically marked as proxied or absent in the user's time table.
> 
> We can also think about some algorithm to give approx % of attendance (if possible)
> 
> Create a Database of colleges, year, course to get approx amount classes taken and then we can provide approx no. of classes he/she can miss (also % of attendance) (there will be 2 % of attendance, one with considering proxy and one without proxy, just to be safe)

But the number of classes in total for any subject also depend upon the teacher which can be changed, I understand that's why approximate, but I believe it could still be a bit misleading. That is why, we will calculate wether or nat the next class can be skipped based on current attendance.

**Comment by Varun-Kolanu** [comment]
The number of classes will be variable and every teacher doesn't announce how many classes they will take in the whole course. If announced too, it can vary depending upon the holidays from the teacher, any emergency in college etc.

So yeah the student will choose the percentage of attendance he needs (Some toppers may try for 90% attendance 😵‍💫). Then according to it, the current attendance percentage will be calculated and the user can try to maintain it regularly. 

We can use the algorithm which we use generally to calculate how many classes more should we attend to reach the desired percentage:

```
((currentAttended + x) / (currentTotal + x)) * 100 = desiredPercentage
```

 and solve for x

**Comment by criticic** [comment]
Still some method of estimating number of classes might be helpful, so we could provid like warnings like "You cannot miss any mmore classes this semester", "7/10 classes missed" which would help someone plan ahead

**Comment by Varun-Kolanu** [comment]
@criticic Yeah nice. If his current percentage is more than desired, we can show how many classes he can miss continuously to not fall below desired and if it's less than desired, we will show how many classes he should attend continuously to reach the desired.

**Comment by devrungta** [comment]
These last two ideas are actually how bunkit works as well..😁😁👍👍

**Comment by Rasesh2005** [comment]
> @devrungta Hehe nice one actually. I forget almost everything 😭
> 
> In attendance tracking, I am too lazy to open the app and mark attendance. So I deleted the app eventually 😅 For this, the following can be done:
> 
> 1. There should be a time, which user choses. At that time daily on working days, app sends a notification to mark attendance.
> 2. We can put a feature that user can save his classroom location. So, whenever he comes inside some radius, with class as center, app sends a notification whether to mark attendance for that class.
> 
> It might be so repetitive to create the time table of the class. So, in the Social Integration section we can add the following features:
> 
> 1. User can see the friends' time table created. Then there will be a button to copy the same time table.
> 2. A user can choose who can see their attendance: Friends, Everyone, No one
> 3. Generally, if we forget to mark whether class was cancelled we will not be able to remember what happened. So, the cancellation of classes will be visible to the friends no matter what, because no privacy is getting disturbed in that. Then, a friend can see the class cancellations from other friends and mark the same in his.
> 4. We can request our friends to mark proxy. Friends can accept or reject the request. Only one user can accept the proxy request. When a friend has accepted a proxy request, all the other requests to friends will be revoked. Then the accepted friend can mark whether he was successful in marking proxy or not. Based on that, the attendance of that day will be automatically marked as proxied or absent in the user's time table.

## Added Feature suggestions:
* Event Addition: Add events such as fest schedule, exam schedule, submission deadlines to their calender(and maybe even the entire class).
* Customization Options: Let users customize the app's theme, layout, and colors according to their preferences.
* Widget Support: Add widget support for quick access to attendance marking and schedule overview on the home screen.
* Student required percentage attendance may vary from course to course( Some course teachers many not require attendance) so we can make this feature individual to every subject.

**Comment by Varun-Kolanu** [comment]
@Rasesh2005 Yeah noicee. We shall add them too

**Comment by criticic** [comment]
came across this app: https://teachassistapp.github.io/

We could also implement features prresent in this app somewhat that is, Marks calculation (from weightage entered for midsem. endsem, assignment etc),
and expected spi calculation, like enter whatever grade you feel you might get and it predicts your spi (and also could help you, say you aim to have more than 8.5 CPI, so you could find out what grades you'd need to achieve it)

also the ui of this app seems nice

**Comment by devrungta** [comment]
We can classify events as academic and non academic, for the academic ones, we can make the option of attendance, grades, credits, and use all academic ones to calculate spi or cpi, classroom(with classmates from contacts), requests for proxy, and then we can add exams in it, location wise attendance marking, 
For non academic we can have event name, attendance( for like fest meets or club meets), time and location attendance marking, and add groups for people working together, for partners and all event time table for all to copy and paste as well.

**Comment by Varun-Kolanu** [comment]
@criticic Pretty useful feature and the app too is nice for reference 🚀

**Comment by Varun-Kolanu** [comment]
We will try to build the full application first using shared_preferences and then if time permits, will move to the friends etc using firebase. Because this is still a good app without the friends' part. Adding firebase in the middle of project might add unnecessary complexity.

**Comment by Varun-Kolanu** [comment]
@Rasesh2005 I was thinking to store jsons in shared_preferences but yeah we can start our initial app too with integrating firebase.
